File: projects/NoizuPromptLingo/src/npl_mcp/npl/parser.py
"""
Expression parser for NPL loading expressions.

Parses expressions like:
- "syntax" -> load entire syntax section
- "syntax#placeholder" -> load specific component
- "syntax#placeholder:+2" -> load with priority filter
- "syntax directives" -> load multiple sections
- "syntax -syntax#literal" -> load with subtraction

Grammar:
    expression      := term (WS term)*
    term            := addition | subtraction
    addition        := section_ref
    subtraction     := '-' section_ref
    section_ref     := section ('#' component)? (':' priority)?
    section         := 'syntax' | 'directives' | 'pumps' | 'prefixes' |
                       'special-sections' | 'declarations' | 'prompt-sections' | 'fences'
    component       := SLUG
    priority        := '+' NUMBER
    SLUG            := [a-z][a-z0-9-]*
    NUMBER          := [0-9]+
    WS              := ' '+
"""

from dataclasses import dataclass
from enum import Enum
from typing import List, Optional
import re

from .exceptions import NPLParseError


class NPLSection(Enum):
    """All loadable NPL sections."""
    SYNTAX = "syntax"
    DECLARATIONS = "declarations"
    DIRECTIVES = "directives"
    PREFIXES = "prefixes"
    PROMPT_SECTIONS = "prompt-sections"
    SPECIAL_SECTIONS = "special-sections"
    PUMPS = "pumps"
    FENCES = "fences"


# Mapping from string names to enum values (supports case-insensitive lookup)
_SECTION_MAP = {s.value.lower(): s for s in NPLSection}
_VALID_SECTIONS = ", ".join(sorted(s.value for s in NPLSection))


@dataclass
class NPLComponent:
    """A parsed NPL component reference."""
    section: NPLSection
    component: Optional[str]  # None means entire section
    priority_max: Optional[int]  # None means all priorities


@dataclass
class NPLExpression:
    """A parsed NPL loading expression."""
    additions: List[NPLComponent]
    subtractions: List[NPLComponent]


# Regex patterns for parsing
_SECTION_REF_PATTERN = re.compile(
    r'^(?P<section>[a-z][a-z0-9-]*)'          # Section name
    r'(?:#(?P<component>[a-z][a-z0-9-]*))?'   # Optional #component
    r'(?::(?P<priority>\+?\-?[0-9]+|\+[a-z]+))?$'  # Optional :priority (capture invalid too)
)


def _parse_section_ref(term: str, is_subtraction: bool = False) -> NPLComponent:
    """Parse a single section reference.

    Args:
        term: The term to parse (e.g., "syntax#placeholder:+2")
        is_subtraction: Whether this is a subtraction term

    Returns:
        NPLComponent with parsed values

    Raises:
        NPLParseError: If term is invalid
    """
    # Check for invalid separators first
    if '@' in term:
        raise NPLParseError(
            f"Invalid expression: '{term}'. Use '#' to separate section from component, "
            f"not '@'. Example: syntax#placeholder"
        )

    # Check for double colons
    if '::' in term:
        raise NPLParseError(
            f"Invalid expression: '{term}'. Found '::' - use single ':' for priority. "
            f"Example: syntax#placeholder:+2"
        )

    match = _SECTION_REF_PATTERN.match(term)
    if not match:
        raise NPLParseError(
            f"Invalid expression: '{term}'. Expected format: section[#component][:+N]. "
            f"Example: syntax#placeholder:+2"
        )

    section_name = match.group('section').lower()
    component = match.group('component')
    priority_str = match.group('priority')

    # Validate section name
    if section_name not in _SECTION_MAP:
        raise NPLParseError(
            f"Unknown section: '{section_name}'. Valid sections: {_VALID_SECTIONS}"
        )

    section = _SECTION_MAP[section_name]

    # Validate and parse priority
    priority_max = None
    if priority_str is not None:
        # Must start with + and be a number
        if not priority_str.startswith('+'):
            raise NPLParseError(
                f"Invalid priority format: '{priority_str}'. Use :+N where N is a non-negative number. "
                f"Example: syntax#placeholder:+2"
            )

        priority_value = priority_str[1:]  # Remove leading +

        # Check if it's a valid number
        if not priority_value.isdigit():
            raise NPLParseError(
                f"Invalid priority format: '+{priority_value}'. Priority must be a non-negative number. "
                f"Example: syntax#placeholder:+2"
            )

        priority_max = int(priority_value)

    return NPLComponent(
        section=section,
        component=component,
        priority_max=priority_max
    )


def parse_expression(expr: str) -> NPLExpression:
    """Parse an NPL loading expression.

    Args:
        expr: The expression to parse

    Returns:
        NPLExpression with additions and subtractions

    Raises:
        NPLParseError: If expression is invalid

    Examples:
        >>> parse_expression("syntax")
        NPLExpression(additions=[NPLComponent(SYNTAX, None, None)], subtractions=[])

        >>> parse_expression("syntax#placeholder:+2")
        NPLExpression(additions=[NPLComponent(SYNTAX, "placeholder", 2)], subtractions=[])

        >>> parse_expression("syntax -syntax#literal")
        NPLExpression(
            additions=[NPLComponent(SYNTAX, None, None)],
            subtractions=[NPLComponent(SYNTAX, "literal", None)]
        )
    """
    # Check for empty or whitespace-only expression
    if not expr or not expr.strip():
        raise NPLParseError("Expression cannot be empty")

    # Split on whitespace
    terms = expr.strip().split()

    additions: List[NPLComponent] = []
    subtractions: List[NPLComponent] = []

    for term in terms:
        if not term:  # Skip empty terms from multiple spaces
            continue

        if term.startswith('-'):
            # Subtraction term
            sub_term = term[1:]  # Remove leading -
            if not sub_term:
                raise NPLParseError(
                    "Invalid subtraction: '-' must be followed by a section reference. "
                    "Example: -syntax#literal"
                )
            component = _parse_section_ref(sub_term, is_subtraction=True)
            subtractions.append(component)
        else:
            # Addition term
            component = _parse_section_ref(term, is_subtraction=False)
            additions.append(component)

    # Must have at least one addition
    if not additions:
        raise NPLParseError(
            "Expression must include at least one section to load. "
            "Cannot only have subtractions."
        )

    return NPLExpression(additions=additions, subtractions=subtractions)

File: lib/noizu_prompt_lingua/npl/parser.ex
defmodule NoizuPromptLingua.NPL.Parser do
  @moduledoc false

  @sections ~w(syntax declarations directives prefixes prompt-sections special-sections pumps fences)
  @section_set MapSet.new(@sections)
  @valid_sections Enum.sort(@sections) |> Enum.join(", ")

  @section_ref_regex ~r/^(?P<section>[a-z][a-z0-9-]*)(?:#(?P<component>[a-z][a-z0-9-]*))?(?::(?P<priority>\+?\-?[0-9]+|\+[a-z]+))?$/

  defmodule Component do
    @moduledoc false
    defstruct [:section, :component, :priority_max]

    @type t :: %__MODULE__{
            section: String.t(),
            component: String.t() | nil,
            priority_max: non_neg_integer() | nil
          }
  end

  defmodule Expression do
    @moduledoc false
    defstruct additions: [], subtractions: []

    @type t :: %__MODULE__{
            additions: [NoizuPromptLingua.NPL.Parser.Component.t()],
            subtractions: [NoizuPromptLingua.NPL.Parser.Component.t()]
          }
  end

  @spec parse(String.t()) :: {:ok, Expression.t()} | {:error, String.t()}
  def parse(expr) when is_binary(expr) do
    trimmed = String.trim(expr)

    if trimmed == "" do
      {:error, "Expression cannot be empty"}
    else
      trimmed
      |> String.split(~r/\s+/, trim: true)
      |> parse_terms([], [])
    end
  end

  defp parse_terms([], additions, subtractions) do
    if additions == [] do
      {:error, "Expression must include at least one section to load. Cannot only have subtractions."}
    else
      {:ok, %Expression{additions: Enum.reverse(additions), subtractions: Enum.reverse(subtractions)}}
    end
  end

  defp parse_terms(["-" <> "" | _rest], _additions, _subtractions) do
    {:error, "Invalid subtraction: '-' must be followed by a section reference. Example: -syntax#literal"}
  end

  defp parse_terms(["-" <> sub_term | rest], additions, subtractions) do
    case parse_section_ref(sub_term) do
      {:ok, component} -> parse_terms(rest, additions, [component | subtractions])
      {:error, _} = err -> err
    end
  end

  defp parse_terms([term | rest], additions, subtractions) do
    case parse_section_ref(term) do
      {:ok, component} -> parse_terms(rest, [component | additions], subtractions)
      {:error, _} = err -> err
    end
  end

  defp parse_section_ref(term) do
    cond do
      String.contains?(term, "@") ->
        {:error,
         "Invalid expression: '#{term}'. Use '#' to separate section from component, not '@'. Example: syntax#placeholder"}

      String.contains?(term, "::") ->
        {:error,
         "Invalid expression: '#{term}'. Found '::' - use single ':' for priority. Example: syntax#placeholder:+2"}

      true ->
        case Regex.named_captures(@section_ref_regex, term) do
          nil ->
            {:error,
             "Invalid expression: '#{term}'. Expected format: section[#component][:+N]. Example: syntax#placeholder:+2"}

          captures ->
            section_name = String.downcase(captures["section"])

            if not MapSet.member?(@section_set, section_name) do
              {:error, "Unknown section: '#{section_name}'. Valid sections: #{@valid_sections}"}
            else
              component = if captures["component"] == "", do: nil, else: captures["component"]

              case parse_priority(captures["priority"]) do
                {:ok, priority_max} ->
                  {:ok,
                   %Component{
                     section: section_name,
                     component: component,
                     priority_max: priority_max
                   }}

                {:error, _} = err ->
                  err
              end
            end
        end
    end
  end

  defp parse_priority(""), do: {:ok, nil}
  defp parse_priority(nil), do: {:ok, nil}

  defp parse_priority("+" <> rest) do
    case Integer.parse(rest) do
      {n, ""} when n >= 0 -> {:ok, n}
      _ -> {:error, "Invalid priority format: '+#{rest}'. Priority must be a non-negative number. Example: syntax#placeholder:+2"}
    end
  end

  defp parse_priority(other) do
    {:error,
     "Invalid priority format: '#{other}'. Use :+N where N is a non-negative number. Example: syntax#placeholder:+2"}
  end
end
